include/fi_atom.h

```diff
@@ -53,150 +53,149 @@ extern "C" {
-	atomic_store(&atomic->val, value);
-	return value;
-}
-
-static inline int atomic_get(atomic_t *atomic)
-{
-	ATOMIC_IS_INITIALIZED(atomic);
-	return atomic_load(&atomic->val);
-}
-
-/* avoid using "atomic_init" so we don't conflict with symbol/macro from stdatomic.h */
-static inline void atomic_initialize(atomic_t *atomic, int value)
-{
-	atomic_init(&atomic->val, value);
-#if ENABLE_DEBUG
-	atomic->is_initialized = 1;
-#endif
-}
-
-static inline int atomic_add(atomic_t *atomic, int val)
-{
-	ATOMIC_IS_INITIALIZED(atomic);
-	return atomic_fetch_add_explicit(&atomic->val,
-			val, memory_order_acq_rel) + 1;
-}
-
-static inline int atomic_sub(atomic_t *atomic, int val)
-{
-	ATOMIC_IS_INITIALIZED(atomic);
-	return atomic_fetch_sub_explicit(&atomic->val,
-			val, memory_order_acq_rel) - 1;
-}
+	typedef atomic_int	atomic_int32_t;
+	typedef atomic_llong	atomic_int64_t;
+
+#define OFI_ATOMIC_DEFINE(radix)									\
+	typedef struct {										\
+		atomic_int##radix##_t val;								\
+		ATOMIC_DEF_INIT;									\
+	} atomic##radix##_t;										\
+													\
+	static inline											\
+	int##radix##_t atomic_inc##radix(atomic##radix##_t *atomic)					\
+	{												\
+		ATOMIC_IS_INITIALIZED(atomic);								\
+		return (int##radix##_t)atomic_fetch_add_explicit(&atomic->val, 1,			\
+								 memory_order_acq_rel) + 1;		\
+	}												\
+	static inline											\
+	int##radix##_t atomic_dec##radix(atomic##radix##_t *atomic)					\
+	{												\
+		ATOMIC_IS_INITIALIZED(atomic);								\
+		return (int##radix##_t)atomic_fetch_sub_explicit(&atomic->val, 1,			\
+								 memory_order_acq_rel) - 1;		\
+	}												\
+	static inline											\
+	int##radix##_t atomic_set##radix(atomic##radix##_t *atomic, int##radix##_t value)		\
+	{												\
+		ATOMIC_IS_INITIALIZED(atomic);								\
+		atomic_store(&atomic->val, value);							\
+		return (int##radix##_t)value;								\
+	}												\
+	static inline											\
+	int##radix##_t atomic_get##radix(atomic##radix##_t *atomic)					\
+	{												\
+		ATOMIC_IS_INITIALIZED(atomic);								\
+		return (int##radix##_t)atomic_load(&atomic->val);					\
+	}												\
+	static inline											\
+	void atomic_initialize##radix(atomic##radix##_t *atomic, int##radix##_t value)			\
+	{												\
+		atomic_init(&atomic->val, value);							\
+		ATOMIC_INIT(atomic);									\
+	}												\
+	static inline											\
+	int##radix##_t atomic_add##radix(atomic##radix##_t *atomic, int##radix##_t val)			\
+	{												\
+		ATOMIC_IS_INITIALIZED(atomic);								\
+		return (int##radix##_t)atomic_fetch_add_explicit(&atomic->val, val,			\
+								 memory_order_acq_rel) + 1;		\
+	}												\
+	static inline											\
+	int##radix##_t atomic_sub##radix(atomic##radix##_t *atomic, int##radix##_t val)			\
+	{												\
+		ATOMIC_IS_INITIALIZED(atomic);								\
+		return (int##radix##_t)atomic_fetch_sub_explicit(&atomic->val, val,			\
+								 memory_order_acq_rel) - 1;		\
```

Comment:
It looks like most calling sites ignore the return code.  I'm hoping the compiler drops the extra math in such cases.  In other uses, the site is wanting the updated value (not surprising, since that was the prior definition).  So, I think these changes are fine.